File: spiral_matrix.py
class Solution:
    def spiralOrder(self,matrix):
        if not matrix:
            return []
        direct=0
        up=left=0
        m=len(matrix)
        n=len(matrix[0])
        down=m-1
        right=n-1
        res=[]
        while True:
            if direct==0:
                for x in range(left,right+1):
                    res.append(matrix[up][x])
                up+=1
            if direct==1:
                for x in range(up,down+1):
                    res.append(matrix[x][right])
                right-=1
            if direct==2:
                for x in range(right,left-1,-1):
                    res.append(matrix[down][x])
                down-=1
            if direct==3:
                for x in range(down,up-1,-1):
                    res.append(matrix[x][left])
                left+=1
            if len(res)==(n*m):
                return res
            direct=(direct+1)%4
        
    def spiralOrder1(self, matrix):
        if not matrix:
            return []
        m = len(matrix)
        n = len(matrix[0])
        row, col = 0, -1
        res = []
        while 1:
            for i in range(n):
                col +=1
                res.append(matrix[row][col]) 
            m -=1
            if m == 0:
                return res
            for j in range(m):
                row +=1
                res.append(matrix[row][col])
            n -=1
            if n==0:
                return res
            for i in range(n):
                col -=1
                res.append(matrix[row][col])
            m -= 1
            if m ==0:
                return res
            for j in range(m):
                row -= 1
                res.append(matrix[row][col])
            n -=1
            if n == 0:
                return res   
if __name__=='__main__':
    sol=Solution()
    matrix=[[ 1, 2, 3 ], [ 4, 5, 6 ], [ 7, 8, 9 ]]
    print sol.spiralOrder(matrix)

File: Algorithm/CPP/spiral_matrix.cpp
# include <iostream>
# include <string>
using namespace std;



class Solution {
public:
	vector<int> spiralOrder(vector<vector<int>>& matrix) {
		int m = matrix.size();
		vector<int> res;
		if (m < 1) return res;
		int n = matrix[0].size();
		int count = m*n;
		int up = 0;
		int down = m - 1;
		int left = 0;
		int right = n - 1;
		while (count)
		{
			for (int j = left; j <= right; j++)
			{
				res.push_back(matrix[up][j]);
				count--;
			}
			up++;
			if (count == 0) return res;
			for (int i = up; i <= down; i++)
			{
				res.push_back(matrix[i][right]);
				count--;
			}
			right--;
			if (count == 0) return res;
			for (int j = right; j >= left; j--)
			{
				res.push_back(matrix[down][j]);
				count--;
			}
			down--;
			if (count == 0) return res;
			for (int i = down; i >= up; i--)
			{
				res.push_back(matrix[i][left]);
				count--;
			}
			left++;
			if (count == 0) return res;
		}
		return res;
	}
};
int main()
{
	Solution sol;
	vector<int> nums = { 100, 4, 200, 1, 3, 2 };//4
	//nums = { 0, 0, -1 };//2
	//nums = { 1, 2, 0, 1 };//3
	vector<vector<int>> matrix = {
		{1,2,3},
		{4,5,6},
		{7,8,9}
	};
	matrix = { { 2, 3 } };
	for(auto k:sol.spiralOrder(matrix)) cout<<k<<" ";
	cout << endl;


	system("pause");
	return 0;

}
